pending-checkmarks-cf-worker/DESIGN.md

```diff
@@ -0,0 +1,71 @@
+# Design spec
+
+Author: [@NoahSaso](https://github.com/NoahSaso)
+
+### KV Mappings
+
+Key | Value
+-- | --
+`WALLET_FOR:<pending_session_id>` | `<wallet_address>`
+`SESSION_FOR:<wallet_address>` | `<pending_session_id>`
+`CURRENT_SESSION_FOR:<initial_session_id>` | `<current_session_id>`
+`SESSION_SEEN:<session_id>` | `1`
+
+### Flows
+1. person verifies for the first time
+	1. get unique session ID from synaps (`pending_session_id`), and store two KV store keys: (1) mapping `pending_session_id` and `wallet_address`, and (2) mapping `wallet_address` to `pending_session_id`
+	2. show pending status
+2. person checks their status by connecting their wallet
+	1. if they have a checkmark assigned, display status
+	2. after a button and a wallet signature request, retrieve `pending_session_id` from KV store using `wallet_address`
+	3. query synaps for status and display it
+		- either pending or failure; if successful, the checkmark should have been assigned
+4. synaps session verification updates and sends a webhook
+	- on uniqueness requirement failure:
+		1. retrieve `current_session_id` using the `initial_session_id` retrieved from the uniqueness requirement failure which provides the initially successful session ID
+		2. check if a checkmark exists with `checkmark_id` set to the sha-512 hash of the `current_session_id`
+		3. if exists, do nothing, leaving the failed `pending_session_id` in the KV store so the wallet can view the failure reason. a uniqueness failed session found in the KV store indicates that a checkmark is already assigned
+		4. if checkmark does not exist:
+			1. assign a checkmark to the wallet, retrieving `wallet_address` from the KV store using `pending_session_id`, with `checkmark_id` set to the sha-512 hash of `pending_session_id`
+			2. map `initial_session_id` to `pending_session_id`, replacing `current_session_id`, to indicate that this is the new session for which the checkmark is currently assigned associated with that first initial verified session
+			3. delete both `wallet_address` <=> `pending_session_id` mappings as they are no longer needed
```

Comment:
So does this mean if I register with a wallet, and somehow am in our system, but don't have a checkmark....? What does this pertain to/when may we hit this case?

I'm a little confused here. I think it's assigning a checkmark during failure that confuses me. Perhaps some additional commentary here would help.

Reply:
yeah, the spec definitely needs a more sophisticated rewrite, this is more a distillation of the logic that will be applied.

the key here is: only one session will succeed for a given unique person. all future verifications will fail due to uniqueness (regardless of checkmark status) and provide the first successful session ID. we then check if a checkmark exists and provide one if not, and do nothing otherwise.